clients/src/main/java/org/apache/kafka/common/network/SslTransportLayer.java

```diff
@@ -267,15 +268,25 @@ public void handshake() throws IOException {
         } catch (IOException e) {
             maybeThrowSslAuthenticationException();
 
-            // this exception could be due to a write. If there is data available to unwrap,
-            // process the data so that any SSL handshake exceptions are reported
-            if (handshakeStatus == HandshakeStatus.NEED_UNWRAP && netReadBuffer.position() > 0) {
-                try {
+            // This exception could be due to a write. If there is data available to unwrap in the buffer, or data available
+            // in the socket channel to read and unwrap, process the data so that any SSL handshake exceptions are reported.
+            try {
+                boolean unwrappedAvailable;
+                do {
+                    int position = netReadBuffer.position();
                     handshakeUnwrap(false);
-                } catch (SSLException e1) {
-                    maybeProcessHandshakeFailure(e1, false, e);
-                }
+                    // handshakeUnwrap unwraps data in the buffer, looping until handshakeStatus != NEED_UNWRAP. Here we
+                    // want to process all incoming data regardless of handshakeStatus. If some data was unwrapped, but
+                    // there is more remaining, try again while `handshakeUnwrap` is able to unwrap any data.
+                    unwrappedAvailable = position != netReadBuffer.position() && netReadBuffer.position() != 0;
```

Comment:
Would it be simpler if we passed a parameter to `handshakeUnwrap` to control whether the handshake status was checked or not?

Reply:
@ijuma Thanks for the review, updated.